### PR TITLE
Refactor Charts for Table-like (x,y) Inputs

### PR DESCRIPTION
**Pull request by rothnic**
This utilizes the existing builder/chart structure, but pretty much replaces the guts of it.

**Summary of approach:**
* Replaces DataAdapter with ChartDataSource, which wraps the pandas dataframe
* Converts multiple input types into pandas dataframe
* Building the glyphs is built around a modified pandas groupby process, which yields DataGroups, which will roughly correspond to one composite glyph where that makes sense. The DataGroup is a subset of data, with the attributes that are assigned to it.
* Includes method for describing the attributes (color, marker, etc.) in a way that users can easily modify
* Builders specify their dimensions (default=(x,y)), so there is opportunity to support charts with something other than x,y
* In general, Builders will work at the entire DataFrame level, and custom glyph behaviors, calculations, etc will be factored out into a very simple and deterministic creation of a composite glyph with more simple data types (array-like data, single attribute (one color, one marker, etc.)).

**Background**
* For a detailed walkthrough of the experimentation that went into this approach see [this notebook](http://nbviewer.ipython.org/gist/rothnic/a6424b0179bd9b29405b/assign_attrs.ipynb).
* For the research that went into the API, see [the bokeh days wiki page](https://github.com/bokeh/bokeh/wiki/Bokeh-Days-Working-Document).
* For the issue associated with the API update, see bokeh/bokeh#2442.

Once complete, this will close bokeh/bokeh#2442. Currently, this is very much still in work, but things should go more quickly from here.

### CODE REVIEW

**Comment by rothnic** [comment]
Screenshot of the output of [this updated scatter example](https://github.com/bokeh/bokeh/blob/ed5275a52d1becac3b15fade5311622ac80b70bb/examples/charts/file/scatter.py).

image

**Comment by bryevdv** [comment]
Looks like tests are failing for `izip` import:
```
bokeh/charts/tests/test_builder.py:20: in <module>
    from bokeh.charts._builder import Builder
bokeh/charts/__init__.py:3: in <module>
    from .builder.area_builder import Area
bokeh/charts/builder/area_builder.py:29: in <module>
    from .._builder import Builder, create_and_build
bokeh/charts/_builder.py:22: in <module>
    from ._data_source import ChartDataSource
bokeh/charts/_data_source.py:20: in <module>
    from itertools import islice, product, izip
E   ImportError: cannot import name izip
``` 
I think there is a py2/3 compat zip in `six`

**Comment by datnamer** [comment]
Looking forward to this! Is it possible to use this ChartDataSource with one disk or distributed blaze table objects? Is that even feasible?

**Comment by rothnic** [comment]
@datnamer: I haven't yet implemented the blaze support, but I have the place to implement it specified. Initial support would be simple, in that it would just be converting to a dataframe internally, so the out of core charting wouldn't be available. It would make it simple to support quick exploration of database-like structures though.

We have decided to utilize pandas, but not depend on blaze due to it still maturing at this point. It is possible that if there are few enough data manipulation operations, we could request them of the chart data source, which would use pandas by default, but if given a blaze resource and blaze was available, it would use blaze instead.

Current plan is to get the charts refactored to support the named column API first, then investigate more advanced internals after that.

**Comment by rothnic** [comment]
Worked through a fair bit more of the internal functionality, but this still could use some more work. The main updates is that there is now an initial implementation of chart properties with automatic error handling for "bad" input configurations. Also, refactored some of Chart and Builder because choosing the axis type is dependent on knowing what the range types are provided to the chart from the builders.

Now scatter can support a single column input:

image

**Comment by datnamer** [comment]
Understood, thanks.

**Comment by rothnic** [comment]
Added legends back in. Really need a better way to specify the mapping between attribute and distinct column value in a better way. I didn't see a way to do anything with legend besides having string: glyph. 

image

**Comment by rothnic** [comment]
Initial bar chart example. I realized while doing this both that the index/cat/labels field is really the same thing as stacking/grouping. It is just a special case where there is a single categorical array used. In this example, 'cyl' is used. However, if I was to use ['cyl', 'origin'], this would produce a nested relationship, which would be able to be displayed with stacking and/or grouping.

The initial bar chart implementation is the first example that includes some kind of aggregation, which required some additional thought on how this would be approached. The way I implemented it was to provide a potential path towards aggregating on the javascript side in the future.

image

**Comment by bryevdv** [comment]
At a glance it looks like the majority of test failures are related to `izip`, which can be fixed up by using 
```
from six.moves import zip
```
instead (this `zip` is a generator)

**Comment by rothnic** [comment]
More work on bar charts, examples included now are:

image

**Comment by rothnic** [comment]
More bar functionality:

image

**Comment by bryevdv** [comment]
Nice. So, do you think enough will be far along enough try to merge some preliminary work for `0.10` (end of September) or hold out for  `0.11` release?

**Comment by rothnic** [comment]
I think end of September is pretty likely, shooting for mid September to get some people playing around with it. I'd like to have the most used charts pretty well implemented by then, and the basic functionality for the remaining.

A lot of the work so far has been around the generalization of some of the concepts, so once i get a few more things worked out, the rest won't be so bad. The core functionality supports taking discrete values (dimgray example above), one or more column labels, or a function that returns a specification for producing unique attributes.

```python
color='blue'
color='cyl'
color=['cyl', 'origin']
color=color(do special things here)
```

Bar is probably one of, if not the most involved, and I just got stacking in there in a way which will support the functional style of use in the future (`stack(Bar(1), Bar(2), Bar(3)`). It isn't quite ready for that use case, but the foundation is there for it. I currently am applying a stack operation to all the BarGlyphs internally. 

The next thing I have left for it is grouping, and then I think area, histogram, and boxplot will be able to leverage some common functionality.

**Comment by rothnic** [comment]
Some rework to utilize as much overlapping functionality between components.

* Bar is now just a specialized Interval (interval is specified in grammar of graphics)
    * will support other composite glyphs, such as bar, and maybe a timeline/gantt-like chart
* BoxGlyph added, which is composted of two Intervals, a ScatterGlyph, and some one-off glyphrenderers to support the whiskers
* The BoxPlot builder subclasses the BarBuilder. The box plot is 95% identical to the bar plot.
* Scatter refactored into composite glyph, with defaults, etc.
* Initial Stats models are added, which provides a future method of serializing the aggregation specifications so the aggregation can happen on the javascript side. 
* Made ScatterBuilder support categorical-like data and used it to add a dot plot-like example (see below).
* Added some default alpha so scatter looks better with many points overlapping and increased default size a bit.

**Initial Box Example**:
image

**Updated Scatter Example**:
image

**Comment by bryevdv** [comment]
FYI python 3 stuff
```
bryan@0199-bryanv (git:charts) ~/work/bokeh/examples/charts/file $ python boxplot.py 
Traceback (most recent call last):
  File "boxplot.py", line 5, in <module>
    from bokeh.charts import BoxPlot, output_file, show
  File "/Users/bryan/work/bokeh/bokeh/charts/__init__.py", line 8, in <module>
    from ._attributes import AttrSpec, color
  File "/Users/bryan/work/bokeh/bokeh/charts/_attributes.py", line 136, in <module>
    class MarkerAttr(AttrSpec):
  File "/Users/bryan/work/bokeh/bokeh/charts/_attributes.py", line 138, in MarkerAttr
    iterable = List(String, default=marker_types.keys())
  File "/Users/bryan/work/bokeh/bokeh/properties.py", line 627, in __init__
    super(List, self).__init__(item_type, default=default, help=help)
  File "/Users/bryan/work/bokeh/bokeh/properties.py", line 594, in __init__
    super(Seq, self).__init__(default=default, help=help)
  File "/Users/bryan/work/bokeh/bokeh/properties.py", line 124, in __init__
    self.validate(default)
  File "/Users/bryan/work/bokeh/bokeh/properties.py", line 605, in validate
    raise ValueError("expected an element of %s, got %r" % (self, value))
ValueError: expected an element of List(String), got KeysView(OrderedDict([('circle', <class 'bokeh.models.markers.Circle'>), ('square', <class 'bokeh.models.markers.Square'>), ('triangle', <class 'bokeh.models.markers.Triangle'>), ('diamond', <class 'bokeh.models.markers.Diamond'>), ('inverted_triangle', <class 'bokeh.models.markers.InvertedTriangle'>), ('asterisk', <class 'bokeh.models.markers.Asterisk'>), ('cross', <class 'bokeh.models.markers.Cross'>), ('x', <class 'bokeh.models.markers.X'>), ('circle_cross', <class 'bokeh.models.markers.CircleCross'>), ('circle_x', <class 'bokeh.models.markers.CircleX'>), ('square_x', <class 'bokeh.models.markers.SquareX'>), ('square_cross', <class 'bokeh.models.markers.SquareCross'>), ('diamond_cross', <class 'bokeh.models.markers.DiamondCross'>)]))
```
Just needs wrapping in `list` probably

**Comment by bryevdv** [comment]
BTW if you close this PR and make a new on in the main Bokeh repo it makes collaboration a bit easier (I could have just fixed that issue, e.g)

**Comment by bryevdv** [comment]
Here's a patch that gets `scatter`, `bar`, and `boxplot` working on py3 (may be other things to fix up)
```
diff --git a/bokeh/charts/_attributes.py b/bokeh/charts/_attributes.py
index c233364..6a4a5a9 100644
--- a/bokeh/charts/_attributes.py
+++ b/bokeh/charts/_attributes.py
@@ -135,7 +135,7 @@ class ColorAttr(AttrSpec):
 
 class MarkerAttr(AttrSpec):
     name = 'marker'
-    iterable = List(String, default=marker_types.keys())
+    iterable = List(String, default=list(marker_types.keys()))
 
     def __init__(self, **kwargs):
         iterable = kwargs.pop('markers', None)
diff --git a/bokeh/charts/_chart.py b/bokeh/charts/_chart.py
index 602e89e..4ca7442 100644
--- a/bokeh/charts/_chart.py
+++ b/bokeh/charts/_chart.py
@@ -22,6 +22,7 @@ from __future__ import absolute_import
 
 import numpy as np
 from collections import defaultdict
+from six import iteritems
 
 from bokeh.charts import ChartOptions, defaults
 from ..browserlib import view
@@ -87,7 +88,7 @@ class Chart(Plot):
 
         # sets overridden defaults
         # ToDo: allow Chart/Plot properties as well as ChartOptions
-        for option, value in option_props.iteritems():
+        for option, value in iteritems(option_props):
             if value != default_props[option]:
                 setattr(self._options, option, value)
 
diff --git a/bokeh/charts/_data_source.py b/bokeh/charts/_data_source.py
index 444c422..65ca416 100644
--- a/bokeh/charts/_data_source.py
+++ b/bokeh/charts/_data_source.py
@@ -15,6 +15,7 @@ methods.
 
 from __future__ import absolute_import
 
+from six import iteritems
 from six.moves import zip
 from operator import itemgetter
 from itertools import islice, product
@@ -90,7 +91,7 @@ def groupby(df, **specs):
         for name, data in df.groupby(spec_cols):
 
             attrs = {}
-            for spec_name, spec in specs.iteritems():
+            for spec_name, spec in iteritems(specs):
                 if spec.columns is not None:
                     # get index of the unique column values grouped on for this spec
                     name_idx = tuple([spec_cols.index(col) for col in spec.columns])
@@ -114,7 +115,7 @@ def groupby(df, **specs):
     # collect up the defaults from the attribute specs
     else:
         attrs = {}
-        for spec_name, spec in specs.iteritems():
+        for spec_name, spec in iteritems(specs):
             attrs[spec_name] = spec[None]
 
         yield DataGroup(label='all', data=df, attr_specs=attrs)
@@ -290,7 +291,7 @@ class ChartDataSource(object):
 
         required_dims = self._required_dims
         selections = self._selections
-        dims = [dim for dim, sel in selections.iteritems() if sel is not None]
+        dims = [dim for dim, sel in iteritems(selections) if sel is not None]
 
         # look for a match for selections to dimensional requirements
         if len(required_dims) > 0:
@@ -311,7 +312,7 @@ class ChartDataSource(object):
                         '\n\nAvailable columns are: %s'
             req_str = [' and '.join(['%s = <Any Column>' % dim for dim in required_dim])
                        for required_dim in required_dims]
-            selection_str = ['%s = %s' % (str(dim), str(sel)) for dim, sel in selections.iteritems() if sel is not None]
+            selection_str = ['%s = %s' % (str(dim), str(sel)) for dim, sel in iteritems(selections) if sel is not None]
 
             raise ValueError(error_str % (' or '.join(req_str), ', '.join(selection_str), ', '.join(self.columns)))
         else:
diff --git a/bokeh/charts/glyphs.py b/bokeh/charts/glyphs.py
index 675308a..8cedcff 100644
--- a/bokeh/charts/glyphs.py
+++ b/bokeh/charts/glyphs.py
@@ -2,6 +2,7 @@ from __future__ import absolute_import
 
 from collections import defaultdict
 import numpy as np
+from six import iteritems
 
 from bokeh.properties import Float, String, Datetime, Bool, Instance
 from bokeh.enums import Aggregation
@@ -102,7 +103,7 @@ class AggregateGlyph(CompositeGlyph):
             filtered_glyphs = self.filter_glyphs(glyphs)
             grouped = self.groupby(filtered_glyphs, 'label')
 
-            for index, group in grouped.iteritems():
+            for index, group in iteritems(grouped):
                 group = sorted(group, key=lambda x: x.stack_label)
                 shift = []
                 for i, glyph in enumerate(group):
@@ -122,7 +123,7 @@ class AggregateGlyph(CompositeGlyph):
             width = min(0.2, (1. / len(grouped.keys())) ** 1.1)
 
             # set bar attributes and re-aggregate
-            for i, (index, group) in enumerate(grouped.iteritems()):
+            for i, (index, group) in enumerate(iteritems(grouped)):
                 for glyph in group:
                     glyph.dodge_shift = step[i + 1]
                     glyph.width = width
diff --git a/bokeh/charts/utils.py b/bokeh/charts/utils.py
index 8256c34..cd213d6 100644
--- a/bokeh/charts/utils.py
+++ b/bokeh/charts/utils.py
@@ -19,6 +19,7 @@ import itertools
 from math import cos, sin
 import pandas as pd
 from pandas.io.json import json_normalize
+from six import iteritems
 
 from ..browserlib import view
 from ..document import Document
@@ -244,7 +245,7 @@ def ordered_set(iterable):
 
 def collect_attribute_columns(**specs):
     """Collect list of unique and ordered columns across attribute specifications."""
-    selected_specs = {spec_name: spec for spec_name, spec in specs.iteritems() if spec.columns}
+    selected_specs = {spec_name: spec for spec_name, spec in iteritems(specs) if spec.columns}
     return ordered_set(list(itertools.chain.from_iterable([spec.columns for spec in selected_specs.values()])))
 
 
@@ -254,7 +255,7 @@ def df_from_json(data, **kwargs):
     if isinstance(data, list):
         return json_normalize(data, kwargs)
     elif isinstance(data, dict):
-        for k, v in data.iteritems():
+        for k, v in iteritems(data):
             if isinstance(v, list):
                 return json_normalize(v)
 
```

**Comment by rothnic** [comment]
@bryevdv: i'm going to take a long pass through when I get to a stopping point for python 3 compatibility and updated docs.

At this point I could tidy up Scatter, Bar, Box, and Histogram, but the rest of the charts will remain broken for now. We can get those in, or we can wait for the next release and I'll have them all ready at that point. Let me know what is preferred.

Here is histogram with auto bin width calculation, and showing different input options:
image

**Comment by rothnic** [comment]
Line example, and line handling datetime values.

image

**Comment by fpliger** [comment]
@rothnic sorry for MIA here.. I went through the examples and the code, looks really promising.

**Comment by rothnic** [comment]
@fpliger: thanks, I'll let you know here soon when it is worth taking a hard look at it. I have some of the main chart types implemented, with the most major feature missing being handling rotation for the aggregating charts (bar, box, histogram). I'm planning on Friday to take a good pass through to clean things up, handle remaining python3 issues, and make sure everything has docstrings.

For the most part, the Chart class has been left as is. The previous implementation handled data transforms, grouping, and creating glyph renders within the builder. Now, the builders are more just handling the assignment of attributes to groups, any unique range building, and different "modes" of operation, then the composite glyphs implement the special graphics behavior for building glyph renderers and transforming them when stacked/dodged. Generally, builders work with table-like data, and composite glyphs work on series/column/array-like data.

Small update, dash attribute added to line. Need to add a special input so you don't need to list out each measure over and over, but here it is:

image

**Comment by bryevdv** [comment]
@rothnic this is great, let's plan to make an up/down decision on including this in `0.10` or not early next week. Basically I'm personally OK with a little breakage if the benefit is high enough. @fpliger @tonyfast any inputs or comments you can provide before then are welcome.

**Comment by rothnic** [comment]
Some additional histogram features for coloring and specifying manual bins:

image

**Comment by rothnic** [comment]
Didn't get as much done this weekend as I hoped. Going to work some on cleaning things up tomorrow in the current state so it could be merged if desired. I'll be less available towards mid-week through the weekend, since I'll be moving.

Added an example in scatter with the `blend` operation I added. This is implemented as specified in grammar of graphics and helps address the problem of column-oriented, timeseries-like, wide data. In this, I combine the width and length Iris columns from 4 to 2 columns, one on x and the other on y:

image

**Comment by rothnic** [comment]
I need to finish the tests, but this should bring things up to date and make the tests pass. I commented out the chart tests that I haven't implemented yet, and have placeholders for the builders I have implemented.

**Comment by bryevdv** [comment]
@rothnic I think there are a couple of problems causing a cascade of test failures.

**Comment by rothnic** [comment]
@bryevdv: I'll get it fixed up early today, then will work on expanding the tests and addressing the documentation.

**Comment by bryevdv** [comment]
I've marked it as `ready`. @rothnic can you summarize what, if anything, is actively broken as a result of this PR? @fpliger and @tonyfast and @bokeh/core please check this branch out and leave comments. We should decd tomorrow afternoon (CST) whether to merge for `0.10` or not.

**Comment by birdsarah** [comment]
Hi, I don't have anything profound, but have downloaded and run this. The merge conflict is because of my recent changes to charts.builder.tests - its' very easy to resolve and inconsequential.

Here are my thoughts:

* This is awesome, am loving things like `scatter6 = Scatter(flowers, x=blend('petal_length', 'sepal_length', name='length'), y=blend('petal_width', 'sepal_width', name='width'), color='species')`
* I want to see this go into 0.10, and it to be the new charts
* That said, I think there are enough charts that don't work that we should make some kind of concessions. My proposal is that we copy the current charts into charts_legacy. People's charts may break if they upgrade to 0.10, but it'll be as simple as changing their import to charts_legacy with appropriate warnings accompanying.

* The following charts.builder.tests have been changed to `pass`
````python
./test_area_builder.py:        pass
./test_donut_builder.py:        pass
./test_dot_builder.py:        pass
./test_heatmap_builder.py:        pass
./test_horizon_builder.py:        pass
./test_step_builder.py:        pass
./test_timeseries_builder.py:        pass
````
I would much rather that we have proper tests, that will fail, and decorate them as `pytest.xfail`. That better reflects our intentions.

* I have run through all the examples in examples.charts.file - these are my findings.
````
area.py - not working - chart not ready
bar.py
blaze_input.py - not working - should work?
boxplot.py - should output `boxplot.html`
cat_heatmap.py - not working - chart not ready
donut.py - not working - chart not ready
dots.py - not working - chart not ready
heatmap.py - not working - chart not ready
histograms.py
horizon.py - not working - chart not ready
horizon_folds.py - not working - chart not ready
hover_span.py - not working - should work?
iris_scatter.py - not working - should work?
lines.py
palettes.py - not working - chart not ready (uses Area)
scatter.py
stacked_bar.py - not working - should work?
steps.py - not working - chart not ready 
stocks_timeseries.py - not working - chart not ready 
scatter.py
````
I think all the examples should work. Those that could work should be updated, those that don't should import from `charts_legacy` (as per my above proposal).

* in the not too distant future i might want to move `charts.conftest` my experience with pytest fixtures is that it can be better to have them all in one place. but this is not for the PR. just noting it so I remember.

**Comment by rothnic** [comment]
* Documentation is broken and docstrings need to be updated. I'd take care of that over the next week or so.
* Currently only implements bar, boxplot, scatter, line, histogram. Line can do timeseries, but we could keep the name with some special defaults and potential additions in the future.
* Yes, tests are stubbed out. Those are currently in work locally.
* At the moment it will be a bit rough beyond the examples I've worked with, but have some ideas on automating tests to apply the large number of potential input conditions they will see.

**Comment by bryevdv** [comment]
> That said, I think there are enough charts that don't work that we should make some kind of concessions. My proposal is that we copy the current charts into charts_legacy. People's charts may break if they upgrade to 0.10, but it'll be as simple as changing their import to charts_legacy with appropriate warnings accompanying.

Is there a reason we could not also leave imports for legacy charts (like `Area`, that don't have new versions) in `bokeh.charts`? Then existing users of `Area`, `Horizon` would not have to change anything for the time being.

**Comment by fpliger** [comment]
>That said, I think there are enough charts that don't work that we should make some kind of concessions. My proposal is that we copy the current charts into charts_legacy. People's charts may break if they upgrade to 0.10, but it'll be as simple as changing their import to charts_legacy with appropriate warnings accompanying.

>> Is there a reason we could not also leave imports for legacy charts (like Area, that don't have new versions) in bokeh.charts? Then existing users of Area, Horizon would not have to change anything for the time being.

I'd be +1 on that... Otherwise I think we should not stretch this into `0.10`. I think the  `charts_legacy` approach could create confusion and would break existing code anyway.

**Comment by bryevdv** [comment]
Here are things that could happen in other PRs, including after/during the freeze, up to about next Tuesday, in my opinion (just my opinion, need others to chime in, especially @damianavila):
* documentation
* clean up tests
* add `charts_legacy`
* add transitive imports from `charts_legacy` where appropriate (`Horizon`, `Area`, etc)
* clean up/prune examples
* a new `TimeSeries`, ***if*** it is just a thin wrapper around `Line` with some different defaults 

Do people have thoughts, opinions on this? If this is acceptable, do we agree to merge this PR today after clearing the conflict?

**Comment by bryevdv** [comment]
> I'd be +1 on that... Otherwise I think we should not stretch this into 0.10. I think the charts_legacy approach could create confusion and would break existing code anyway.

I elaborated a bit more in an edit @fpliger, I was suggesting doing both, basically. Do you agree/disagree with that? We could name it `_charts_legacy` to be clear, and add warning to any usage there.

**Comment by birdsarah** [comment]
> Is there a reason we could not also leave imports for legacy charts (like Area, that don't have new versions) in bokeh.charts? Then existing users of Area, Horizon would not have to change anything for the time being. 

I'm not sure, but my assumption was that you need to change your charts code to work with the new charts builder. So, for example, people who were importing from `Bar` and don't want to immediately refactor their chart code can just import `charts_legacy.Bar`

**Comment by birdsarah** [comment]
> add transitive imports from charts_legacy where appropriate (Horizon, Area, etc)

oh yes, I agree with that.

**Comment by bryevdv** [comment]
Cool just to be more clear/succinct, my proposal is:
* move *all* old chart code to `bokeh._charts_legacy`
* add warnings to everything in `bokeh._charts_legacy`
* add selected transitive imports to `bokeh.charts` for old charts that don't yet have new counterparts

**Comment by fpliger** [comment]
> I elaborated a bit more in an edit @fpliger, I was suggesting doing both, basically. Do you agree/disagree with that? We could name it _charts_legacy to be clear, and add warning to any usage there.

Agree :+1: . That would mostly make things transparent to the normal user and would give us the opportunity to gradually replace the other "missing" charts in the next minor releases.

**Comment by birdsarah** [comment]
> Here are things that could happen in other PRs, including after/during the freeze, up to about next Tuesday, in my opinion (just my opinion, need others to chime in, especially @damianavila):
* documentation
* clean up tests
* add charts_legacy
* add transitive imports from charts_legacy where appropriate (Horizon, Area, etc)
* clean up/prune examples
* a new TimeSeries, if it is just a thin wrapper around Line with some different defaults

I think we should seperate the must-haves from the nice-to-haves.

I would say `charts_legacy`  and `clean up examples` are must-have. 

And the `nice-to-haves` would be in a priority order e.g.
* clean up tests
* documentation
* transitive imports
* new TimeSeries

**Comment by birdsarah** [comment]
> Cool just to be more clear/succinct, my proposal is:
* move *all* old chart code to `bokeh._charts_legacy`
* add warnings to everything in `bokeh._charts_legacy`
* add selected transitive imports to `bokeh.charts` for old charts that don't yet have new counterparts

+1

**Comment by bryevdv** [comment]
OK, @birdsarah do you think yo can clear the merge conflict? Then we could merge this afternoon.

**Comment by birdsarah** [comment]
I don't have push to rothnic's repo.

**Comment by bryevdv** [comment]
Ah, this is why PR branches of the main repo are to be preferred. If you click on `command line instructions` in the merge UI below it shows how you can fix and merge the PR using a manual merge on the command line.

**Comment by damianavila** [comment]
My gut feeling is that we are merging this too soon... if we merge this we have to do a plenty of work against the clock to ship something half done... I understand that everyone want to ship great things as soon as possible, but, at overall I am :-1: to merge this one in this context...

**Comment by birdsarah** [comment]
but I have pushed up the conflict free version to our version.

**Comment by birdsarah** [comment]
@bryevdv https://github.com/bokeh/bokeh/tree/rothnic-feature/xy_charts_api

**Comment by bryevdv** [comment]
@damianavila OK well we need to come to a decision then, let's let some others weigh in. My own feeling is that even rushed and half-done this still represents a clear improvement. I think it is of paramount importance to get this more "R-like" syntax out in front of people, because on several occasions now folks have said they wouldn't consider using `bokeh.charts`, even more importantly, wouldn't consider *teaching* with `bokeh.charts`, *specifically* because of the cumbersome syntax.

**Comment by birdsarah** [comment]
one additional TODO, one of the examples downloaded some sampledata into sampledata folder - which isn't how we handle sample data.

**Comment by bryevdv** [comment]
@damianavila an additional observation: all of this is self-contained in `bokeh/charts` and `bokeh/_legacy_charts` if things somehow go terribly south after a merge we can simply swap everything back in the way it was. I suppose my position is that merging this now gives us an opportunity to get to including it in the release. Waiting now will preclude that, I think.

**Comment by fpliger** [comment]
Just to be clear. I don't think it's going to be painless and foresee things broken... but charts is openly not a stable API. But if I have to balance between waiting 0.11 for the new charts to come in and taking the risk of breaking (some) charts in 0.10 (to be fixed in the next minor release) I'd choose the second option.

**Comment by damianavila** [comment]
@bokeh/core I think I am the only one :-1: and because this is a decision by the group... if several of you think it is OK to do it now, well, let's do it... 
I just wanted to make clear my position about this :wink:

**Comment by birdsarah** [comment]
Superceded by a branch on bokeh repo - #2870.
